### ISSUE TITLE
Null argument issue with ArgumentActionHolder

### ISSUE THREAD

**Issue by seregvan** [appccelerate/statemachine#15]
There is a problem with ArgumentActionHolder. We can't provide null as argument because of missing check for null. Now

`if (argument != Missing.Value && !(argument is T))`

should be:

`if (argument != Missing.Value && argument != null && !(argument is T))`

**Comment by seregvan** [comment]
By the way, ArgumentGuardHolder already has this check.

**Comment by ursenzler** [comment]
Thanks for spotting this issue.

I'll fix it as soon as possible. Or if you are in a hurry: fork the project, fix it, open a pull request and you can get an updated pre-release nuget package from https://www.myget.org/gallery/appccelerate (look for a pre-release package that was just added).

**Comment by seregvan** [comment]
Thank you! :)

**Comment by ursenzler** [comment]
I took a closer look and null can be passed as an argument.

Can you better describe what goes wrong in your situation?

Passing `null` works because the value to be passed to the action is initialized with `default(T)` and the checks are ignored.

**Comment by seregvan** [comment]
In my code I used something like 
`stateMachine.In(...).On(SomeEvent.GotoEvent1).Goto(...).Execute<SomeClass>(...)`
and when I use
`stateMachine.Fire(SomeEvent.GotoEvent1, null) `
I get an exception that argument can't be converted (just from ArgumentActionHolder.Execute).
So for me the problem is that (argument is T) is not working in this case, because argument is null.
And as I can see, default(T) is used for castArgument local variable, but first check with ArgumentException   applied to argument method parameter.
In my case I fixed this by adding argument != null check and everything works fine now.
Where am I wrong?

**Comment by ursenzler** [comment]
Hmm

This test code runs (on my machine):
```
                var m = new PassiveStateMachine<string, string>();

                FooBar value = new FooBar("initial");

                m.In("A")
                    .On("X").Goto("B").Execute((FooBar fb) => { value = fb; });

                m.Initialize("A");

                m.Start();

                m.Fire("X", null);

                value.Should().BeNull();
```
                
I start wondering about the version you are running.

**Comment by ursenzler** [comment]
Okay, the debugger is confusing me. 
My tests runs successfully but when I debug, the values shown make currently no sense (because of the `argument is T`, which should work as you explain but it doesn't?!?!?!?

Will investigate tomorrow...

**Comment by seregvan** [comment]
I made test too

```
internal class TestArgument
{
    public TestArgument(int num) { Number = num; }
    public int Number { get; private set; }
}
...
var stateMachine = new PassiveStateMachine<string, string>();
var ta = new TestArgument(1);
stateMachine.In("A").On("X").Goto("B").Execute<TestArgument>(arg => ta = arg);

stateMachine.Initialize("A");
stateMachine.Start();
stateMachine.Fire("X", null);

Assert.AreEqual(null, ta);
```

When I have this in ArgumentActionHolder

`if (argument != Missing.Value && !(argument is T))`

Test failed with exception

Test method AppccelerateTests.UnitTest1.TestArgumentActionHolder threw exception: 
Appccelerate.StateMachine.Machine.StateMachineException: No exception listener is registered. Exception:  ---> System.ArgumentException: Cannot cast argument to match action method. Argument = , Action = <TestArgumentActionHolder>b__0
    at Appccelerate.StateMachine.Machine.ActionHolders.ArgumentActionHolder`1.Execute(Object argument) in D:\Work\RainyDay\Current\Appcelerate\statemachine\Appccelerate.StateMachine\Machine\ActionHolders\ArgumentActionHolder.cs:line 43
   at Appccelerate.StateMachine.Machine.Transitions.Transition`2.PerformActions(ITransitionContext`2 context) in D:\Work\RainyDay\Current\Appcelerate\statemachine\Appccelerate.StateMachine\Machine\Transitions\Transition.cs:line 225

But when I change to

`if (argument != Missing.Value && argument != null && !(argument is T))`

Then Test is passed.
I'm using current sources from github (I got them around a week ago).

**Comment by ursenzler** [comment]
I found that on my machine with VS2017 evaluates `null != Missing.Value` to `false`. Which should not be in my understanding and as I can remember was not the case when I wrote the state machine (.net version, compiler version).
Could that be a breaking change in the .net Framework/Compiler/what-ever??

So I absolutely agree to add the check for `null` as you propose, but I'm confused why the value of `Missing.Value` equals `null` now.

I'll add the check for null and publish a new package, anyways.

**Comment by seregvan** [comment]
Hm... As I understand Missing.Value is not null, it's an object for use like a marker of the missing value. It can't be null because null is a value too. So I don't really understand why you are using this. So maybe you need to change first if to

`if ((argument == Missing.Value) || (argument != null && !(argument is T)))`

and then the exception will be thrown when argument equals Missing.Value (that is the way you can undestand that there are no value at all and you can't cast this argument) or value is not null and not of Type you are looking for.

And there are no cases left for direct cast to be failed and you can change

```
if (argument != Missing.Value)
{
    castArgument = (T)argument;
}
this.action(castArgument);
```

just to 

`this.action((T)argument);`

and there will be no problems. Less code, easier to undestand.

**Comment by ursenzler** [comment]
Latest version contains a fix for this issue. Thanks again for pointing out the issue.

**Comment by ursenzler** [comment]
fixed in version 3.2.0
